File: Zig/src/service/user_service.zig
const std = @import("std");
const UserRepository = @import("../repository/user_repository.zig").UserRepository;
const User = @import("../models/user.zig").User;
const CreateUserRequest = @import("../models/user.zig").CreateUserRequest;
const UpdateUserRequest = @import("../models/user.zig").UpdateUserRequest;

pub const UserService = struct {
    repo: *UserRepository,
    allocator: std.mem.Allocator,

    pub fn init(repo: *UserRepository, allocator: std.mem.Allocator) UserService {
        return UserService{
            .repo = repo,
            .allocator = allocator,
        };
    }

    pub fn createUser(self: *UserService, req: CreateUserRequest) !User {
        // Validation could go here
        return self.repo.create(req.name, req.email, self.allocator);
    }

    pub fn getAllUsers(self: *UserService) ![]User {
        return self.repo.getAll(self.allocator);
    }

    pub fn getUserById(self: *UserService, id: i64) !?User {
        return self.repo.getById(id, self.allocator);
    }

    pub fn updateUser(self: *UserService, id: i64, req: UpdateUserRequest) !?User {
        return self.repo.update(id, req.name, req.email, self.allocator);
    }

    pub fn deleteUser(self: *UserService, id: i64) !bool {
        return self.repo.delete(id);
    }
};

File: Rust/src/service/user_service.rs
use crate::models::user::{CreateUser, UpdateUser, User};
use crate::repository::user_repository::UserRepository;
use anyhow::Result;

#[derive(Clone)]
pub struct UserService {
    repo: UserRepository,
}

impl UserService {
    pub fn new(repo: UserRepository) -> Self {
        Self { repo }
    }

    pub async fn create_user(&self, payload: CreateUser) -> Result<User> {
        self.repo.create(payload).await
    }

    pub async fn get_all_users(&self) -> Result<Vec<User>> {
        self.repo.find_all().await
    }

    pub async fn get_user_by_id(&self, id: i64) -> Result<Option<User>> {
        self.repo.find_by_id(id).await
    }

    pub async fn update_user(&self, id: i64, payload: UpdateUser) -> Result<Option<User>> {
        self.repo.update(id, payload).await
    }

    pub async fn delete_user(&self, id: i64) -> Result<bool> {
        self.repo.delete(id).await
    }
}
